voter-legacy/index.md

```diff
@@ -0,0 +1,739 @@
+
+In between `<script></script>` tags below this
+
+```js
+export default {
+  props: {
+    type: {
+      default: "p1",
+    },
+  },
+};
+```
+
+Now in our `App.vue` we need to update to fetch our votes. In the `<script></script>` tag we should have the following resulting code:
+
+### `vue/src/App.vue`
+
+```js
+export default {
+  created() {
+    this.$store.dispatch("cosmos/init");
+    this.$store.dispatch("cosmos/entityFetch", {type: "poll", module: "voter"});
+    this.$store.dispatch("cosmos/entityFetch", {type: "vote", module: "voter"});
+  },
+};
+```
+
+By now should be able to see the same UI as in the first screenshot. Try creating polls and casting votes. You may notice that it's possible to cast multiple votes for one poll. This is not what we want, so let's fix this behaviour.
+
+## Casting votes only once
+
+To fix this issue we first have to understand how data is stored in our application.
+
+We can think of our data storage as a lexicographically ordered key value store. You can loop through the entries, filter by key prefix, add, update and delete entries. It is easier to visualize the store as JSON:
+
+```json
+{
+  "poll-1a266241-c58d-4cbc-bacf-aaf939c95de1": {
+    "creator": "cosmos15c6g4v5yvq0hy3kjllyr9ytlx45r936y0m6dm6",
+    "id": "1a266241-c58d-4cbc-bacf-aaf939c95de1",
+    "title": "Soft drinks",
+    "options": ["Coca-Cola", "Pepsi"]
+  },
+  "vote-cd63b110-2959-45b0-8ce3-afa2fb7a5652": {
+    "creator": "cosmos15c6g4v5yvq0hy3kjllyr9ytlx45r936y0m6dm6",
+    "id": "cd63b110-2959-45b0-8ce3-afa2fb7a5652",
+    "pollID": "1a266241-c58d-4cbc-bacf-aaf939c95de1",
+    "value": "Pepsi"
+  }
+}
+```
+
+Both `poll-` and `vote-` are prefixes. They are added to keys for ease of filtering. By convention, prefixes are defined in `x/voter/types/key.go`.
+
+### `x/voter/keeper/vote.go`
+
+Whenever a user casts a vote, a new "create vote" message is handled by a handler and is passed to a keeper. Keeper takes a `vote-` prefix, adds a UUID (unique to every message) and uses this string as a key. `x/voter/keeper/vote.go`:
+
+```go
+key := []byte(types.VotePrefix + vote.ID)
+```
+
+These strings are unique and we get duplicate votes. To fix that we need to make sure a keeper records every vote only once by choosing the right key. In our case, we can use poll ID and creator address to make sure that one user can cast only one vote per poll.
+
+```go
+key := []byte(types.VotePrefix + vote.PollID + "-" + string(vote.Creator))
+```
+
+Restart the application and try voting multiple times on a single poll, you'll see you can vote as many times as you want but only your most recent vote is counted.
+
+## Introducing a fee for creating polls
+
+Let's make it so that creating a poll costs 200 tokens.
+
+This feature is very easy to add. We already require users to have accounts registered, and each user has tokens on balance. The only thing we need to do is to send coins from user's account to a module account before we create a poll.
```

Comment:
```suggestion
Now make it so that creating a poll costs 200 tokens.

This feature is very easy to add. You already require users to have accounts registered and each user has tokens on balance. The only thing you need to do is to send coins from the user's account to a module account before we create a poll.
```